test/units/module_utils/test_api.py

```diff
@@ -46,3 +45,53 @@ def login_database():
 
         with pytest.raises(Exception):
             login_database()
+
+    def test_no_retries(self):
+
+        @retry()
+        def login_database():
+            assert False, "Should not execute"
+
+        login_database()
+
+
+class TestRetryWithDelaysAndCondition:
+
+    def test_empty_retry_iterator(self):
+        @retry_with_delays_and_condition(backoff_iterator=[])
+        def login_database():
+            login_database.counter += 1
+
+        login_database.counter = 0
+        r = login_database()
+        assert login_database.counter == 1
+
+    def test_no_retry_exception(self):
+        @retry_with_delays_and_condition(
+            backoff_iterator=[1],
+            should_retry_error=lambda x: False,
+        )
+        def login_database():
+            login_database.counter += 1
+            if login_database.counter == 1:
+                raise Exception("Error")
+
+        login_database.counter = 0
+        with pytest.raises(Exception):
+            login_database()
+        assert login_database.counter == 1
+
+    def test_retry_exception(self):
+        @retry_with_delays_and_condition(
+            backoff_iterator=[1],
+            should_retry_error=lambda x: isinstance(x, Exception),
```

Comment:
It'd be interesting to see a test case for something inheriting a `BaseException` too.